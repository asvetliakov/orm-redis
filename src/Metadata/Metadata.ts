/**
 * Metadata key to mark class to be stored as redis hash
 */
export const REDIS_HASH = Symbol("Redis Hash");

/**
 * Metadata key to store redis properties metadatas
 */
export const REDIS_PROPERTIES = Symbol("Redis properties");

/**
 * Metadata key to keep serialized initial value of property from redis. For collections this will be link to the corresponding hashmap/set
 */
export const REDIS_VALUE = Symbol("Redis initial value");

/**
 * Metadata key to keep the initial state of redis collections (map/set)
 */
export const REDIS_COLLECTION_VALUE = Symbol("Redis initial collection value");


export type PropertyMetadata = SimplePropertyMetadata | RelationPropertyMetadata;

/**
 * Hash simple property metadata
 * 
 * @export
 * @interface SimplePropertyMetadata
 */
export interface SimplePropertyMetadata {
    /**
     * Property name in class
     */
    propertyName: string;
    /**
     * Property name stored in redis
     */
    propertyRedisName: string;
    /**
     * Property type
     */
    propertyType: any;
    /**
     * True if property must be identify column
     */
    isIdentifyColumn: boolean;
    /**
     * Relation flag
     */
    isRelation: false;
}

/**
 * Relation options
 * 
 * @export
 * @interface RelationOptions
 */
export interface RelationOptions {
    /**
     * Property name in redis
     */
    propertyName?: string;
    /**
     * True to automatically create new hash for relation
     */
    cascadeInsert?: boolean;
    /**
     * True to automatically update corresponding relation hash
     */
    cascadeUpdate?: boolean;
}

/**
 * Hash relation property metadata
 * 
 * @export
 * @interface RelationPropertyMetadata
 */
export interface RelationPropertyMetadata {
    /**
     * Property name in class
     */
    propertyName: string;
    /**
     * Property name stored in redis
     */
    propertyRedisName: string;
    /**
     * Property type.
     * For set property relation it will be Set itself
     * For single property relation it will be equal to relationType
     */
    propertyType: any;
    /**
     * Relation flag
     */
    isRelation: true;
    /**
     * Relation type. Must be type of related entity
     */
    relationType: any;
    /**
     * Relation options
     */
    relationOptions: RelationOptions;
}

/**
 * Return true if given object is redis hash like
 * 
 * @export
 * @param hash 
 * @returns 
 */
export function isRedisHash(hash: object): boolean {
    if (typeof hash !== "object" || hash === null) {
        return false;
    }
    const metadata = Reflect.getMetadata(REDIS_HASH, hash.constructor);
    return !!metadata;
}

/**
 * Return redis hash name
 * 
 * @export
 * @param hashOrHashClass 
 * @returns 
 */
export function getRedisHashName(hashOrHashClass: object | Function): string | undefined {
    const metadata = Reflect.getMetadata(REDIS_HASH, typeof hashOrHashClass === "function" ? hashOrHashClass : hashOrHashClass.constructor);
    return metadata;
}

/**
 * Return redis hash id
 * 
 * @export
 * @param hash 
 * @returns 
 */
export function getRedisHashId(hash: { [key: string]: any }): string | number | undefined {
    if (!isRedisHash(hash)) {
        return undefined;
    }
    const metadatas: PropertyMetadata[]  = Reflect.getMetadata(REDIS_PROPERTIES, hash.constructor);
    const idMetadata = metadatas && metadatas.find(val => !val.isRelation && val.isIdentifyColumn);
    if (!idMetadata) {
        return undefined;
    }
    if (typeof hash[idMetadata.propertyName] === "number" || typeof hash[idMetadata.propertyName] === "string") {
        return hash[idMetadata.propertyName];
    }
    return undefined;
}

/**
 * Returns redis properties metadata
 * 
 * @export
 * @param hashOrHashClass 
 * @returns 
 */
export function getRedisHashProperties(hashOrHashClass: object | Function): PropertyMetadata[] | undefined {
    return Reflect.getMetadata(REDIS_PROPERTIES, typeof hashOrHashClass === "function" ? hashOrHashClass : hashOrHashClass.constructor);
}