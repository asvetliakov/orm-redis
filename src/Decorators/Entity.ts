import { REDIS_ENTITY } from "../Metadata/Metadata";

/**
 * Defines redis entity class
 * 
 * @export
 * @param {string} [name] 
 * @returns void
 */
export function Entity(name?: string): ClassDecorator {
    return function (constructor: Function) {
        Reflect.defineMetadata(REDIS_ENTITY, name ? name : constructor.name, constructor);
    };
}