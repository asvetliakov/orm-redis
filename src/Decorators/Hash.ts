import { REDIS_HASH } from "../Metadata/Metadata";

/**
 * Defines redis hash
 * 
 * @export
 * @param {string} [name] 
 * @returns void
 */
export function Hash(name?: string): ClassDecorator {
    return function (constructor: Function) {
        Reflect.defineMetadata(REDIS_HASH, name ? name : constructor.name, constructor);
    };
}