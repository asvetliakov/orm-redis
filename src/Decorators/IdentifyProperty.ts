import { PropertyMetadata, REDIS_PROPERTIES } from "../Metadata/Metadata";

/**
 * Defines identify property for given redis hash.
 * Hash must contain one IdentifyProperty
 * 
 * @export
 * @param type Type
 * @param [name] Redis property name
 * @returns 
 */
export function IdentifyProperty(nameOrType?: string | typeof Number | typeof String, type?: typeof Number | typeof String): PropertyDecorator {
    return function (target: Object, propertyKey: string): void {
        const designType: any = Reflect.getMetadata("design:type", target, propertyKey);
        const name = typeof nameOrType === "string" ? nameOrType : propertyKey;
        const propertyType = typeof nameOrType === "function"
            ? nameOrType
            : type
                ? type
                : designType;        

        if (propertyType !== String && propertyType !== Number) {
            throw new Error("Identify property must be string or number type");
        }
        const properties: PropertyMetadata[] = Reflect.getMetadata(REDIS_PROPERTIES, target.constructor) || [];
        properties.push({
            propertyName: propertyKey,
            propertyRedisName: name,
            isIdentifyColumn: true,
            propertyType: propertyType,
            isRelation: false
        });
        Reflect.defineMetadata(REDIS_PROPERTIES, properties, target.constructor);
    };
}