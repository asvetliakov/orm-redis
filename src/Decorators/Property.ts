import { LazyMap } from "../collections/LazyMap";
import { LazySet } from "../collections/LazySet";
import { PropertyMetadata, REDIS_PROPERTIES } from "../Metadata/Metadata";

export type ValidType =
    typeof Number |
    typeof String |
    typeof Boolean |
    typeof Object |
    typeof Array |
    typeof Date |
    typeof Map |
    typeof LazyMap |
    typeof LazySet |
    typeof Set;


/**
 * Defines redis hash property
 * 
 * @export
 * @param [type] 
 * @returns 
 */
export function Property(type?: ValidType): PropertyDecorator;
/**
 * Defines redis hash property
 * 
 * @export
 * @param name Redis name for property
 * @param [type] 
 * @returns 
 */
export function Property(name: string, type?: ValidType): PropertyDecorator; 

/**
 * Defines redis hash property
 * 
 * @export
 * @param [name] Property name to store in redis
 * @returns 
 */
export function Property(nameOrType?: string | ValidType, type?: ValidType): PropertyDecorator {
    return function (target: Object, propertyKey: string): void {
        const designType: any = Reflect.getMetadata("design:type", target, propertyKey);
        const name = typeof nameOrType === "string" ? nameOrType : propertyKey;
        const propertyType = typeof nameOrType === "function"
            ? nameOrType
            : type
                ? type
                : designType;        

        const properties: PropertyMetadata[] = Reflect.getMetadata(REDIS_PROPERTIES, target.constructor) || [];
        properties.push({
            propertyName: propertyKey,
            propertyRedisName: name,
            isIdentifyColumn: false,
            propertyType: propertyType,
            isRelation: false,
        });
        Reflect.defineMetadata(REDIS_PROPERTIES, properties, target.constructor);
    };
}