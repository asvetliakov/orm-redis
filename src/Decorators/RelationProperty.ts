import { MetadataError } from "../Errors/Errors";
import { PropertyMetadata, REDIS_PROPERTIES, RelationOptions, RelationTypeFunc } from "../Metadata/Metadata";

/** 
 * default options
 */
const defaultOptions: RelationOptions = {
    cascadeInsert: false,
    cascadeUpdate: false
};

/**
 * Defines single or multiple relation property
 * 
 * @export
 * @param type Type function which must return type of relation / property type
 * @param propertyType Property type. For single relation must be 
 */
export function RelationProperty(type: RelationTypeFunc): PropertyDecorator;
/**
 * Defines single or multiple relation property
 * 
 * @export
 * @param type Type function which must return type of relation / property type
 * @param [options] Options
 */
export function RelationProperty(type: RelationTypeFunc, options?: RelationOptions): PropertyDecorator;

/**
 * Defines single or multiple relation property
 * 
 * @export
 * @param type 
 * @param [options] 
 * @returns 
 */
export function RelationProperty(type: RelationTypeFunc, options?: RelationOptions): PropertyDecorator {
    return function (target: Object, propertyKey: string): void {
        const designType = Reflect.getMetadata("design:type", target, propertyKey);
        
        const redisPropertyName = options && options.propertyName ? options.propertyName : propertyKey;
        
        const relationTypes = type();

        let propertyType: Function;
        if (Array.isArray(relationTypes)) {
            [, propertyType] = relationTypes;
        } else {
            if (designType === Object) {
                throw new MetadataError(target.constructor, `Relation's ${propertyKey} property type is detected as simple Object. This is error. Specify property type explicitly`);
            }
            propertyType = designType;
        }

        const finalOptions = {
            ...defaultOptions,
            ...options
        };

        const properties: PropertyMetadata[] = Reflect.getMetadata(REDIS_PROPERTIES, target.constructor) || [];
        properties.push({
            isRelation: true,
            propertyName: propertyKey,
            propertyRedisName: redisPropertyName,
            propertyType: propertyType,
            relationTypeFunc: type,
            relationOptions: finalOptions
        });
        Reflect.defineMetadata(REDIS_PROPERTIES, properties, target.constructor);
    };
}