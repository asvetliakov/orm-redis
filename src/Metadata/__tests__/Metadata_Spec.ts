import { Hash } from "../../Decorators/Hash";
import { IdentifyProperty } from "../../Decorators/IdentifyProperty";
import { Property } from "../../Decorators/Property";
import { getRedisHashId, getRedisHashName, isRedisHash } from "../Metadata";

describe("isRedisHash()", () => {
    it("returns false for non objects", () => {
        expect(isRedisHash(null as any)).toBeFalsy();
        expect(isRedisHash(undefined as any)).toBeFalsy();
        expect(isRedisHash(0 as any)).toBeFalsy();
        expect(isRedisHash("" as any)).toBeFalsy();
        expect(isRedisHash(Symbol() as any)).toBeFalsy();
        expect(isRedisHash(true as any)).toBeFalsy();
    });

    it("returns false non non redis hash classes instances and simple objects", () => {
        class A { }
        expect(isRedisHash(new A())).toBeFalsy();
        expect(isRedisHash({})).toBeFalsy();
    });

    it("returns true for redis hash class instances", () => {
        @Hash()
        class A { }
        expect(isRedisHash(new A())).toBeTruthy();
    });
});

describe("getRedisHashId()", () => {
    it("returns undefined for non redis hashes", () => {
        class A { }
        expect(getRedisHashId(new A())).toBeUndefined();
        expect(getRedisHashId({})).toBeUndefined();
        expect(getRedisHashId(null as any)).toBeUndefined();
    });

    it("returns undefined if redis hash doesn't contain identify property", () => {
        @Hash()
        class A {
            @Property(String)
            public prop: string = "a";
        }
        expect(getRedisHashId(new A())).toBeUndefined();
    });

    it("returns undefined if identify property is not string or number", () => {
        @Hash()
        class A {
            @IdentifyProperty(String)
            public prop: string = new Date() as any;
        }
        expect(getRedisHashId(new A())).toBeUndefined();
    });

    it("returns id", () => {
        @Hash()
        class A {
            @IdentifyProperty(String)
            public prop: string = "abc";
        }
        @Hash()
        class B {
            @IdentifyProperty(Number)
            public prop: number = 5;
        }
        expect(getRedisHashId(new A())).toBe("abc");
        expect(getRedisHashId(new B())).toBe(5);
    });
});

describe("getRedisHashName()", () => {
    it("Returns undefined for non hashes", () => {
        class A { }
        expect(getRedisHashName(new A())).toBeUndefined();
        expect(getRedisHashName(A)).toBeUndefined();
        expect(getRedisHashName({})).toBeUndefined();
    });

    it("Returns redis hash name either for class or object", () => {
        @Hash()
        class A { }

        @Hash("test")
        class B { }
        expect(getRedisHashName(new A())).toBe("A");
        expect(getRedisHashName(A)).toBe("A");
        expect(getRedisHashName(new B())).toBe("test");
        expect(getRedisHashName(B)).toBe("test");
    });
});