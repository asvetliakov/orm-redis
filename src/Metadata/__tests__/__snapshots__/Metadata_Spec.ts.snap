// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`getRedisHashProperties() Returns properties for even hash instance or hash constructor 1`] = `
Array [
  Object {
    "isIdentifyColumn": true,
    "isRelation": false,
    "propertyName": "id",
    "propertyRedisName": "id",
    "propertyType": [Function],
  },
  Object {
    "isIdentifyColumn": false,
    "isRelation": false,
    "propertyName": "prop2",
    "propertyRedisName": "prop2",
    "propertyType": [Function],
  },
]
`;

exports[`getRedisHashProperties() Returns properties for even hash instance or hash constructor 2`] = `
Array [
  Object {
    "isIdentifyColumn": true,
    "isRelation": false,
    "propertyName": "id",
    "propertyRedisName": "id",
    "propertyType": [Function],
  },
  Object {
    "isIdentifyColumn": false,
    "isRelation": false,
    "propertyName": "prop2",
    "propertyRedisName": "prop2",
    "propertyType": [Function],
  },
]
`;
